File: tsp-java/tsp-adapter-html/src/main/java/dev/tsp/html/HtmlAdapter.java
package dev.tsp.html;

import dev.tsp.core.model.*;

public final class HtmlAdapter {

    public static String render(TIRResult tir) {
        StringBuilder sb = new StringBuilder();
        sb.append("<!-- tsp template_id=").append(tir.templateId())
          .append(" schema_id=").append(tir.schemaId()).append(" -->\n");
        for (TIRNode n : tir.nodes()) sb.append(renderNode(n));
        return sb.toString();
    }

    private static String renderNode(TIRNode node) {
        return switch (node) {
            case TIRTextNode t -> t.content();
            case TIRExprNode e -> e.resolved() == null ? "" : escape(String.valueOf(e.resolved()));
            case TIRIfNode i -> {
                StringBuilder s = new StringBuilder();
                for (TIRNode n : i.branch()) s.append(renderNode(n));
                yield s.toString();
            }
            case TIRForeachNode f -> {
                StringBuilder s = new StringBuilder();
                for (java.util.List<TIRNode> item : f.items()) {
                    for (TIRNode n : item) s.append(renderNode(n));
                }
                yield s.toString();
            }
        };
    }

    static String escape(String s) {
        return s
            .replace("&", "&amp;")
            .replace("<", "&lt;")
            .replace(">", "&gt;")
            .replace("\"", "&quot;")
            .replace("'", "&#x27;");
    }

    private HtmlAdapter() {}
}

File: templane-spec/templane-core/src/templane_core/html_adapter.py
from __future__ import annotations
import html as _html_module
from .models import TIRResult, TIRNode, TIRTextNode, TIRExprNode, TIRIfNode, TIRForeachNode


def render(tir: TIRResult) -> str:
    header = f"<!-- templane template_id={tir.template_id} schema_id={tir.schema_id} -->"
    body = "".join(_node(n) for n in tir.nodes)
    return header + "\n" + body


def _node(node: TIRNode) -> str:
    if isinstance(node, TIRTextNode):
        return node.content
    if isinstance(node, TIRExprNode):
        if node.resolved is None:
            return ""
        return _html_module.escape(str(node.resolved))
    if isinstance(node, TIRIfNode):
        return "".join(_node(n) for n in node.branch)
    if isinstance(node, TIRForeachNode):
        return "".join("".join(_node(n) for n in item) for item in node.items)
    return ""
